src/lib.rs

```diff
@@ -182,6 +184,17 @@ pub type Result<T> = result::Result<T, Errno>;
 /// ones.
 pub type Error = Errno;
 
+/// A file descriptor representing the current working directory.
+#[cfg(not(target_os = "redox"))]
+pub const AT_FDCWD: &BorrowedFd<'static> = unsafe {
```

Comment:
> Why

Makes the API clearer IMO. Then you don't need to constantly wrap your FDs with Some.

> For example, there are many functions that can't accept AT_FDCWD.

WDYM? All the old option functions just converted None to AT_FDCWD AFAIK.

Reply:
Ohhh, excellent point! I'm 100% onboard the "express constraint with the type system" bandwagon, but I still don't like using Option and I think this is a great opportunity to revisit this. What about adding a trait (maybe called `IntoPathFd`) implemented for all `AsFd` and a special `Cwd` marker struct. The interface would be `fn as_raw_fd(&self) -> RawFd`. Then open would be `fn open<Fd: IntoPathFd>(fd: Fd, ...)` and write would be `fn write<Fd: AsFd>(fd: Fd, ...)`. `write(Cwd, ...)` would not compile while `open(Cwd, ...)` would. WDYT?

Reply:
Flushed it out a bit, this seems awesome:
```rust
#![allow(dead_code, unused_variables)]
use std::fs::File;
use std::os::unix::prelude::*;

trait IntoPathFd {
    fn as_raw_fd(&self) -> RawFd;
}

impl<Fd: AsFd> IntoPathFd for Fd {
    fn as_raw_fd(&self) -> RawFd {
        AsRawFd::as_raw_fd(&self.as_fd())
    }
}

struct Cwd;

impl IntoPathFd for Cwd {
    fn as_raw_fd(&self) -> RawFd {
        -1
    }
}

fn open(fd: impl IntoPathFd) {}
fn write(fd: impl AsFd) {}

fn foo() {
    open(&File::open("").unwrap());
    open(Cwd);
    // write(Cwd);
    write(&File::open("").unwrap());
}
```

Reply:
> WDYM? All the old option functions just converted None to AT_FDCWD AFAIK.

When the `path` argument is an absolute path, `dirfd` will be ignored. So if I wanna pass an absolute path to those `xxxat()` functions, in the previous implementation:

```rust
xxxat(None, "/absolute/path", ...)
```

What should I pass to the first argument with the new interface:

```rust
fn xxxat<Fd: AsFd>(dirfd: &Fd, ...)
```

`AT_FDCWD`? 
```rust
xxxat(AT_FDCWD, "/absolute/path", ...)
```
I know the value of actual `dirfd` passed into the function does not matter, but the **semantics** has been changed, and becomes a little bit weird to the user IMO...

Reply:
I personally think you should be passing in AT_FDCWD since if your absolute path turns out to be relative, cwd is what's going to be used. That said, with the design I presented above, we could just add a type called `Ignored` or even implement IntoPathFd for options so you can pass in None.